go/vt/vtctl/workflow/server.go

```diff
@@ -3310,19 +3344,30 @@ func (s *Server) switchWrites(ctx context.Context, req *vtctldatapb.WorkflowSwit
 		}
 	}
 
-	// Need to lock both source and target keyspaces.
-	tctx, sourceUnlock, lockErr := sw.lockKeyspace(ctx, ts.SourceKeyspaceName(), "SwitchWrites")
+	// Lock the workflow and its source and target keyspaces.
+	lockName := fmt.Sprintf("%s/%s", ts.TargetKeyspaceName(), ts.WorkflowName())
+	ctx, workflowUnlock, lockErr := s.ts.LockName(ctx, lockName, "SwitchWrites")
+	if lockErr != nil {
+		return handleError(fmt.Sprintf("failed to lock the %s workflow", lockName), lockErr)
+	}
+	defer workflowUnlock(&err)
+	// We need to hold the keyspace locks longer than the command timeout.
+	cmdTimeout := req.GetTimeout()
+	if cmdTimeout == nil {
+		cmdTimeout = &vttimepb.Duration{Seconds: 30}
+	}
+	ksLockTimeout := time.Duration(int64(time.Second) * (cmdTimeout.Seconds * 2))
```

Comment:
Why are we locking for twice the command timeout?

Reply:
The timeout is primarily for the step where we wait for replication to catch up everywhere. So this gives us plenty of time for all of the work being done. I thought about making it even longer since the main thing is that we want to be sure that we do not lose the lock when we're supposed to have exclusive access to the resources we modify.

Reply:
I added some comments to help explain the values used.